src/peft/tuners/vera/model.py

```diff
@@ -0,0 +1,549 @@
+from peft.utils import (
+    TRANSFORMERS_MODELS_TO_VERA_TARGET_MODULES_MAPPING,
+    ModulesToSaveWrapper,
+    _get_submodules,
+)
+
+from .config import VeraConfig
+from .layer import Embedding, Linear, VeraLayer
+
+
+def _vera_forward_hook(module, args, kwargs, vera_A, vera_B):
+    kwargs["vera_A"] = vera_A
+    kwargs["vera_B"] = vera_B
+    return args, kwargs
+
+
+def _kaiming_init(
+    tensor_or_shape: Union[torch.Tensor, tuple],
+    generator: torch.Generator,
+):
+    if isinstance(tensor_or_shape, tuple):
+        tensor = torch.empty(tensor_or_shape)
+    else:
+        tensor = tensor_or_shape
+    fan = _calculate_correct_fan(tensor, "fan_in")
+    gain = math.sqrt(2)
+    std = gain / math.sqrt(fan)
+    bound = math.sqrt(3.0) * std
+
+    with torch.no_grad():
+        return tensor.uniform_(-bound, bound, generator=generator)
+
+
+class VeraModel(BaseTuner):
+    """
+    Creates Vector-based Random Matrix Adaptation (Vera) model from a pretrained transformers model.
+
+    Args:
+        model ([`~transformers.PreTrainedModel`]): The model to be adapted.
+        config ([`VeraConfig`]): The configuration of the Vera model.
+        adapter_name (`str`): The name of the adapter, defaults to `"default"`.
+
+    Returns:
+        `torch.nn.Module`: The Vera model.
+
+    Example:
+
+        ```py
+        >>> from transformers import AutoModelForSeq2SeqLM
+        >>> from peft import VeraModel, VeraConfig
+
+        >>> config = VeraConfig(
+        ...     task_type="SEQ_2_SEQ_LM",
+        ...     r=8,
+        ...     target_modules=["q", "v"],
+        ...     vera_dropout=0.01,
+        ... )
+
+        >>> model = AutoModelForSeq2SeqLM.from_pretrained("t5-base")
+        >>> vera_model = VeraModel(model, config, "default")
+        ```
+
+        ```py
+        >>> import transformers
+        >>> from peft import VeraConfig, PeftModel, get_peft_model
+
+        >>> target_modules = ["q_proj", "k_proj", "v_proj", "out_proj", "fc_in", "fc_out", "wte"]
+        >>> config = VeraConfig(
+        ...     r=4, target_modules=target_modules, vera_dropout=0.1, bias="none", task_type="CAUSAL_LM"
+        ... )
+
+        >>> model = transformers.GPTJForCausalLM.from_pretrained(
+        ...     "kakaobrain/kogpt",
+        ...     revision="KoGPT6B-ryan1.5b-float16",  # or float32 version: revision=KoGPT6B-ryan1.5b
+        ...     pad_token_id=tokenizer.eos_token_id,
+        ...     use_cache=False,
+        ...     device_map={"": rank},
+        ...     torch_dtype=torch.float16,
+        ...     load_in_8bit=True,
+        ... )
+        >>> model = prepare_model_for_int8_training(model)
+        >>> vera_model = get_peft_model(model, config)
+        ```
+
+    **Attributes**:
+        - **model** ([`~transformers.PreTrainedModel`]) -- The model to be adapted.
+        - **peft_config** ([`VeraConfig`]): The configuration of the Vera model.
+    """
+
```

Comment:
Let's add the `prefix` class attribute, similar to what `LoraModel` has.

Reply:
This is trickier than it might appear, as we add both trainable and non-trainable parameters to the model.

What should the prefix be? It could be `vera_` or `vera_lambda_`. The former will include the trainable and non-trainable parameters, whereas the latter just the trainable ones. I am unsure which is more appropriate.

For example https://github.com/huggingface/peft/blob/21c304f6f6ea62c7dcbef8e201d178a7575b471d/src/peft/tuners/lora/model.py#L226C8-L228 sets `requires_grad` to `False` if the prefix is not in the key. However, maybe prefix needs to include non-trainable in other places?

Reply:
I'd go with trainable params only, let's try and see if that causes any problems.

Reply:
Actually, I can see this won't work in advance. See the following block: https://github.com/vvvm23/peft/blob/c22a8e5d47774b3c9ed7352994ccd34e757fff58/src/peft/tuners/lora/model.py#L227-L236

You can see in the first loop that we want to match against the module name using the prefix. But in the second we only want trainable parameters. If we set to `vera_lambda_` (trainable parameters) the first loop will fail. If we set to `vera_` (all parameters) we will wrongly train `vera_A` and `vera_B`.

Proposing `prefix="vera_A"` and `_trainable_prefix="vera_lambda_"`? 🤔

Reply:
Good catch. So I'd still propose to go with `vera_lambda`, i.e. the prefix is only for trainable params. Inside of `_replace_module`, you could still hard-code `vera`, with a comment saying that this cannot use `prefix` because of the mentioned issue.

I think there is a more general solution which would use the `adapter_layer_names` attribute on the adapter layer instead of relying on `prefix`, but `adapter_layer_names` has the same issue of not explicitly distinguishing between trainable vs non-trainable (though the former is assumed). However, I'd leave that for a refactor PR.

Reply:
Okay, in that case the only time we actually use `self.prefix` will be in `_mark_only_adapters_as_trainable`. All others will have a comment saying why we can't use it.